geopandas/tools/sjoin.py

```diff
@@ -186,60 +285,61 @@ def find_contains(a1, a2):
-            .merge(
-                right_df.drop(right_df.geometry.name, axis=1),
-                left_on="_key_right",
-                right_index=True,
-                suffixes=("_%s" % lsuffix, "_%s" % rsuffix),
-            )
-            .set_index(index_left)
-            .drop(["_key_right"], axis=1)
-        )
-        if isinstance(index_left, list):
-            joined.index.names = left_index_name
-        else:
-            joined.index.name = left_index_name
-
-    elif how == "left":
-        result = result.set_index("_key_left")
-        joined = (
-            left_df.merge(result, left_index=True, right_index=True, how="left")
-            .merge(
-                right_df.drop(right_df.geometry.name, axis=1),
-                how="left",
-                left_on="_key_right",
-                right_index=True,
-                suffixes=("_%s" % lsuffix, "_%s" % rsuffix),
-            )
-            .set_index(index_left)
-            .drop(["_key_right"], axis=1)
-        )
-        if isinstance(index_left, list):
-            joined.index.names = left_index_name
-        else:
-            joined.index.name = left_index_name
+    return result
 
-    else:  # how == 'right':
-        joined = (
-            left_df.drop(left_df.geometry.name, axis=1)
-            .merge(
-                result.merge(
-                    right_df, left_on="_key_right", right_index=True, how="right"
-                ),
-                left_index=True,
-                right_on="_key_left",
-                how="right",
+def _sjoin_nearest(left_df,
+    right_df,
+    tolerance,
+    nearest_distances,
+    tree_idx_r,
+    tree_idx_l,
+    ):
+    tree_idx = tree_idx_r or right_df.sindex
+    
+    if tolerance is None:
+        # pick a sensible tolerance
+        rel_tol = int(DEFAULT_NEAREST_REL_TOL * len(right_df) + DEFAULT_NEAREST_ABS_TOL)
+        tolerance = max((rel_tol, len(right_df)))
+        warn(
+                'Using default tolerance. To speed up sjoin when using `op="nearest"`,'
+                ' please specify a tolerance. If your geometries are very close, '
+                ' pick a large tolerance to avoid inconsisitencies.'
             )
-            .set_index(index_right)
-            .drop(["_key_left", "_key_right"], axis=1)
+    else:
+        assert tolerance >= 1 and isinstance(tolerance, (int, long)), "tolerance must be a positive integer"
+    idxmatch = left_df.geometry.apply(lambda x: x.bounds).apply(
+        lambda x: list(tree_idx.nearest(x, tolerance)) if not x == () else []
         )
-        if isinstance(index_right, list):
-            joined.index.names = right_index_name
-        else:
-            joined.index.name = right_index_name
+    if sum(idxmatch.apply(len) > 0):
+        # check each set of matches for actual distance
+        l_idx = []
+        r_idx = []
+        distances = []
+        for ind_in_left, possible_nearest in enumerate(idxmatch):
```

Comment:
This is the iterative part where we take the results from `rtree` and refine them by calculating the distance to each point.